firedrake/dmplex.pyx

```diff
@@ -257,118 +260,149 @@ def closure_ordering(PETSc.DM plex,
+                vertices[off+1 + (3-i)] = c_vertices[i]
+            for i in range(off-1, -1, -1):
+                facets[off-1 - i] = c_facets[i]
+            for i in range(3, off-1, -1):
+                facets[off + (3-i)] = c_facets[i]
+
+        # At this point the cell "has" the right starting vertex
+        # and order of traversal. If the starting vertex is one with an X,
+        # and arrows on the edges show the order of traversal:
         #
-        #   *--<--*
+        #   o--<--o
         #   |     |
         #   v     ^
         #   |     |
-        #   *-->--*
+        #   o-->--X
         #
-        # For outer product elements, we need directions like this:
+        # then outer product elements assume edge directions like this:
         #
-        #   *-->--*
+        #   o--<--o
         #   |     |
         #   ^     ^
         #   |     |
-        #   *-->--*
+        #   o--<--X
         #
-        edge_direction[2] = 1 - edge_direction[2]
-        edge_direction[3] = 1 - edge_direction[3]
-
-        # And edges should be ordered like this:
+        # ... and a vertex ordering like this:
         #
-        #   *--1--*
+        #   3-----1
         #   |     |
-        #   2     3
         #   |     |
-        #   *--0--*
+        #   |     |
+        #   2-----0
         #
-        cell_closure[cell, v_per_cell + 0] = cone[0]
-        cell_closure[cell, v_per_cell + 1] = cone[2]
-        cell_closure[cell, v_per_cell + 2] = cone[3]
-        cell_closure[cell, v_per_cell + 3] = cone[1]
-
-        edge_directions[cell, 0] = edge_direction[0]
-        edge_directions[cell, 1] = edge_direction[2]
-        edge_directions[cell, 2] = edge_direction[3]
-        edge_directions[cell, 3] = edge_direction[1]
-
-        # Vertices should be ordered like this:
+        # ... and a facet (edge) ordering like this:
         #
-        #   2-----3
-        #   |     |
+        #   o--3--o
         #   |     |
+        #   1     0
         #   |     |
-        #   0-----1
+        #   o--2--o
         #
-        bottom, top = cone[0], cone[2]
-        bottom_dir, top_dir = edge_direction[0], edge_direction[2]
-        CHKERR(DMPlexGetCone(plex.dm, bottom, &bottom_vertices))
-        CHKERR(DMPlexGetCone(plex.dm, top, &top_vertices))
-
-        if not bottom_dir:
-            cell_closure[cell, 0] = bottom_vertices[0]
-            cell_closure[cell, 1] = bottom_vertices[1]
-        else:
-            cell_closure[cell, 0] = bottom_vertices[1]
-            cell_closure[cell, 1] = bottom_vertices[0]
+        # So let us permute.
+        cell_closure[cell, 0] = vertices[0]
+        cell_closure[cell, 1] = vertices[1]
+        cell_closure[cell, 2] = vertices[3]
+        cell_closure[cell, 3] = vertices[2]
+        cell_closure[cell, 4 + 0] = facets[0]
+        cell_closure[cell, 4 + 1] = facets[2]
+        cell_closure[cell, 4 + 2] = facets[3]
+        cell_closure[cell, 4 + 3] = facets[1]
+        cell_closure[cell, 8] = c
```

Comment:
OK, so now you rewire the cell closure into the canonical order, the first two vertices are the same and the last two needed swapping; for the facets the dmplex order is just counter-clockwise (following the arrows), hence the permutation.